src/papaline/core.clj

```diff
@@ -239,20 +242,26 @@
   (invoke [this ctx]
           (let [pre-hook  *pre-execution-hook*
                 post-hook *post-execution-hook*
-                result (CompletableFuture.)]
+                result (CompletableFuture.)
+                args-array (if (not-empty (:args ctx))
+                             (.toArray ^clojure.lang.ArraySeq (:args ctx))
+                             (into-array []))]
             (letfn [(clos [stgs ctx]
                       (if-let [^RealizedStage s (first stgs)]
                         (let [[^CompletableFuture next-args-future new-ctx] (run-task2 s ctx error-handler)]
                           (if (.isDone next-args-future)
                             (process-result (assoc new-ctx :args (.get next-args-future)) stgs)
-                            (.handleAsync ^CompletableFuture next-args-future
```

Comment:
previously we use `handleAsync` which may chose any thread available in the pool.